### ISSUE TITLE
ideas for next-generation downloading capability

### ISSUE THREAD

**Issue by rptaylor** [cvmfs/cvmfs#3543]
I had a brief chat with @HereThereBeDragons about parallelizing downloads - not only in the sense of performing multiple downloads concurrently, but also that each download would retrieve data from multiple HTTP servers in parallel.

While the work involved could be significant indeed, so could the benefits:
* It would no longer be necessary to know/care about geolocation (obviating complexities and dependencies on whatever MaxMind feels like doing such as https://github.com/cvmfs/cvmfs/issues/3529). Geolocation is in any case only an estimator attempting to select one 'optimal' s1 server, and potentially a poor estimator as the bandwidth of a link is unrelated to location (though the bandwidth of a stream is certainly related to RTT, that can also depend on TCP buffers and numerous other factors.)
* The concept of failing over from an unreachable stratum server to another one (incurring some delay while it times out) would no longer be applicable.
* No matter whether some servers are unreachable or geographically distant, retrieving data pieces from all servers in parallel and assembling them into the downloaded file should give the theoretically* optimal results; in principle it should be strictly better than (or in the extreme case, equal to) retrieving the data from any one individual server - just download pieces from each server as fast as each server can provide them.

(*) For small files the overhead could make it less beneficial, while for large files it could have more impact.

I got thinking about it a bit more (uh oh) and looked into some options:
1. [aria2](https://github.com/aria2/aria2) This is a CLI client that can download from multiple HTTP servers in parallel using the Metalink standard. To build apps that use this, there is libaria2 however it is [not very well documented](https://github.com/aria2/aria2/blob/master/doc/manual-src/en/libaria2.rst). (That doc says libaria2 is not thread safe but it is also not updated for a long time and may not be applicable anymore). However there have not been very regular releases of aria2 recently, and [libmetalink is not very supported/adopted/maintained](https://daniel.haxx.se/blog/2021/06/07/bye-bye-metalink-in-curl/) anymore, since the torrent protocol won out. (Interesting that libcurl used to have this functionality, but it was removed!)
2. [libtorrent](https://www.libtorrent.org/) "is a feature complete C++ bittorrent implementation focusing on efficiency and scalability. ... It boasts a well documented library interface that is easy to use."
IIUC there is a [WebSeed](http://www.bittorrent.org/beps/bep_0019.html) ([1](https://en.wikipedia.org/wiki/Torrent_file#HTTP_seeds) [2](https://libtorrent.org/reference-Torrent_Info.html#web_seed_entry) ) extension to the torrent protocol which appears to provide the capability to download from multiple specified HTTP servers in parallel. If this capability could be used standalone without other torrent-related complications or requirements it should be possible to use this in the CVMFS client to do parallel downloads.  (There are also [HTTP seeds](http://www.bittorrent.org/beps/bep_0017.html) but this seems different.) This approach could also open the way to future consideration of p2p downloading between CVMFS clients within a private LAN which would be very interesting! (From my point of view I think site admins would be fine with this as long as they can configure and controll it, firewall it on a specific port, etc.) 

Checksumming and decompression can be handled on the fly by some of these options, but how it might integrate/conflict/replace with the way CVMFS does this would need some further investigation. 

I also had a look at [libp2p](https://libp2p.io/) ([docs](https://docs.libp2p.io/), [connectivity](https://connectivity.libp2p.io/)) , which looks like a  comprehensive, thoroughly-documented and widely adopted network stack for p2p applications, but it looks like a bridge too far in the sense of requiring a completely p2p-native architecture and design.

**Comment by DrDaveD** [comment]
How does per-site caching fit into this?

**Comment by rptaylor** [comment]
Assuming you mean forward caching proxies? For pragmatism I think any initial implementation of this idea should avoid any new change or requirements on the overall architecture of proxies + stratum servers. I'm pretty sure both of the options mentioned above (and there could be others) support use of plain http proxies, e.g. http://libtorrent.org/reference-Settings.html#enum-proxy-type-t  so for the most part there would not be a major difference.

However using multiple HTTP sources in parallel would highlight the fact that identical chunks from independent stratum servers are cached (or not) completely separately, which is both unfortunate from a caching efficiency perspective and a vital precept of how HTTP works with location-based rather than content-based addressing (although you can use some clever rules in nginx forward caching proxies to get around this). 

In principle I think proxy sharding should still work with this and would help reduce the proxy disk usage, by a factor of 1/N for N squids, which could be smaller than the number of stratum servers.  Proxies (and disk space for them) are pretty cheap though, even ~100 GB can often be more than enough.

If an implementation reaches a hypothetical 2nd stage of parallelized/torrent-like download capability, where clients in a private network also share data with each other (similar to https://github.com/XenitAB/spegel which is a new and successful idea following this model for container image downloads for containerd), the situation could evolve to a point where eventually traditional caching proxies are not as essential and may not be needed, or possibly something else becomes more useful or efficient (e.g. a seeding peer that acts like something in between a stratum server and a caching proxy).

**Comment by DrDaveD** [comment]
> Assuming you mean forward caching proxies?

Yes, or some equivalent.

> For pragmatism I think any initial implementation of this idea should avoid any new change or requirements on the overall architecture of proxies + stratum servers. I'm pretty sure both of the options mentioned above (and there could be others) support use of plain http proxies, e.g. http://libtorrent.org/reference-Settings.html#enum-proxy-type-t so for the most part there would not be a major difference.

The reference talks about using the `CONNECT` method in an http proxy.  That method is not cachable, so it's not helpful.  It just passes bytes through a separate tcp connection.

> However using multiple HTTP sources in parallel would highlight the fact that identical chunks from independent stratum servers are cached (or not) completely separately, which is both unfortunate from a caching efficiency perspective and a vital precept of how HTTP works with location-based rather than content-based addressing (although you can use some clever rules in nginx forward caching proxies to get around this).

Basically I think in order to do what you're talking about the site caching proxies would need to be involved and be given a lot more smarts.  They would be the ones that would have to connect to multiple servers in a "Happy Eyeballs" type of algorithm which initiates parallel requests and accepts the first response while shutting down the rest.  The biggest challenge with that is getting the configuration to all of them at all the sites and keeping them up to date.   There's no standard mechanism for that which I am aware of; I think it would need to be a custom design, which has a lot of drawbacks, including making it less useful to other applications.

... 
> If an implementation reaches a hypothetical 2nd stage of parallelized/torrent-like download capability, where clients in a private network also share data with each other (similar to https://github.com/XenitAB/spegel which is a new and successful idea following this model for container image downloads for containerd), the situation could evolve to a point where eventually traditional caching proxies are not as essential and may not be needed, or possibly something else becomes more useful or efficient (e.g. a seeding peer that acts like something in between a stratum server and a caching proxy).

I have dreamt of peer-2-peer solutions like that for over 15 years but they all have their limitations and downsides and there has so far at least never been enough motivation to try them.

**Comment by rptaylor** [comment]
> The reference talks about using the `CONNECT` method in an http proxy. That method is not cachable, so it's not helpful. It just passes bytes through a separate tcp connection.

My interpretation of that was "The server is assumed to be an HTTP proxy. If the transport used for the connection is (HTTPS), the server is assumed to support the [CONNECT](http://tools.ietf.org/html/draft-luotonen-web-proxy-tunneling-01) method"
As I understand, CONNECT is the only possible way to access a HTTPS page through a (forward) proxy, so doesn't this make sense in that context? I think this is drawing a distinction between HTTP as the L7 protocol and HTTPS (TLS) as part of the transport layer. And I thought this implies cacheability: "for web seeds and HTTP trackers, a plain proxy will suffice" - i.e. plain as opposed to supporting CONNECT.  

> Basically I think in order to do what you're talking about the site caching proxies would need to be involved and be given a lot more smarts. They would be the ones that would have to connect to multiple servers in a "Happy Eyeballs" type of algorithm which initiates parallel requests and accepts the first response while shutting down the rest. The biggest challenge with that is getting the configuration to all of them at all the sites and keeping them up to date. There's no standard mechanism for that which I am aware of; I think it would need to be a custom design, which has a lot of drawbacks, including making it less useful to other applications.

Supposing that the client simply does multiple GET requests in parallel there should be no major architectural changes required on stratum servers or proxies. The only impact would be using N times more proxy cache disk space when accessing N stratum servers as discussed above, which is slightly suboptimal but probably worth the benefits. But we should confirm whether this premise is correct. What are the technical requirements for a HTTP request from a client in order for a forward proxy to be able to cache it?

**Comment by rptaylor** [comment]
I found a [blog post](https://blog.thelifeofkenneth.com/2019/09/adding-webseed-urls-to-torrent-files.html) showing low level example details of what a torrent file specification that uses regular HTTP servers would look like (using the Webseed extension I mentioned earlier).
The challenge here would be the concept of torrent pieces vs CVMFS chunks. 
A torrent client breaks each file into many pieces, and parallelizes by downloading multiple pieces at a time.
If data file = CVMFS chunk, then torrent pieces means CVMFS sub-chunks. My understanding of BEP 19 (webseeds) is that should be no problem for HTTP servers, they only need to serve the plain data files. However it means that clients would need to have metadata providing checksums with the granularity of pieces (sub-chunks), so if they receive some bad data they can tell which piece needs to be redownloaded to ensure integrity of the whole file (chunk).

From what [I can gather](https://blog.libtorrent.org/2020/09/bittorrent-v2/)  [BEP 30](http://www.bittorrent.org/beps/bep_0030.html) is a clever scheme that would solve this problem, avoiding the need for hashes of all pieces, but it remained in draft status, while torrent file specification v2 adopted a different scheme for checksumming based on Merkle trees, that still seems to require hashes of all pieces in a file (chunk). I don't fully understand this:
```
The leaves of the hash trees are always 16 kiB (the block size), regardless of the piece size. The concept of piece size still exists and is still used in the peer-wire protocol as it is today. e.g. in HAVE and REQUEST messages. However, instead of piece hashes actually being the hash of the content of the piece, it’s the root of the hash tree of the piece. i.e. a sub-tree of the full merkle tree.
```

or how the piece size and block sizes are different.

**Comment by DrDaveD** [comment]
> > The reference talks about using the `CONNECT` method in an http proxy. That method is not cachable, so it's not helpful. It just passes bytes through a separate tcp connection.
> 
> My interpretation of that was "The server is assumed to be an HTTP proxy. If the transport used for the connection is (HTTPS), the server is assumed to support the [CONNECT](http://tools.ietf.org/html/draft-luotonen-web-proxy-tunneling-01) method" As I understand, CONNECT is the only possible way to access a HTTPS page through a (forward) proxy, so doesn't this make sense in that context? I think this is drawing a distinction between HTTP as the L7 protocol and HTTPS (TLS) as part of the transport layer. And I thought this implies cacheability: "for web seeds and HTTP trackers, a plain proxy will suffice" - i.e. plain as opposed to supporting CONNECT.

Yes CONNECT is required to access an https page through a forward proxy.  However it cannot cache the response.  Every https response will be different and only understandable to the client & server that know the encryption key for the connection.  CONNECT only passes bytes through and does not understand them.  That would be a man in the middle attack if it could.

> > Basically I think in order to do what you're talking about the site caching proxies would need to be involved and be given a lot more smarts. They would be the ones that would have to connect to multiple servers in a "Happy Eyeballs" type of algorithm which initiates parallel requests and accepts the first response while shutting down the rest. The biggest challenge with that is getting the configuration to all of them at all the sites and keeping them up to date. There's no standard mechanism for that which I am aware of; I think it would need to be a custom design, which has a lot of drawbacks, including making it less useful to other applications.
> 
> Supposing that the client simply does multiple GET requests in parallel there should be no major architectural changes required on stratum servers or proxies. The only impact would be using N times more proxy cache disk space when accessing N stratum servers as discussed above, which is slightly suboptimal but probably worth the benefits. But we should confirm whether this premise is correct.

I suppose it might potentially work for the clients to do multiple GET requests.  It would also be N times the bandwidth, not just cache space, bandwidth at every step (served by stratum 1s, received by proxies, sent by proxies, received by clients) if the transfers are all allowed to complete.  It would be a major impact on the infrastructure.  There would instead need to be some type of abort on all of the requests that lag, for some definition of lagging.  I don't know what the impact of all those aborts would be.

Before we had the cvmfs geo API, I had come up with a CVMFS bash-based configuration (which OSG used) that tried to do a tcp connection to all the stratum 1s at mount time and select the one that responded the quickest, then use that order for the rest of the mount.  It was quite messy and unreliable, and the geo API design was chosen as a much cleaner replacement of that.

> What are the technical requirements for a HTTP request from a client in order for a forward proxy to be able to cache it?

Basically it needs to be a standard http URL GET request + response.  frontier-squid also puts a limit (by default) on the response size of 1GB.  The standard squid configuration (kept by frontier-squid) will also treat a URL with a question mark ("?") or "cgi-bin" in it which does not supply a cache expiration time in the response as an expiration time of 0 seconds, effectively not caching it.

**Comment by DrDaveD** [comment]
So let's think about only replacing the mount-time geo API lookup with something that would try to dynamically probe the response times via all proxies through to all stratum 1s in parallel.  It would need to read a well-known small file from each stratum 1.  It would need to set `Cache-control: max-age=0` on the request so squid will always forward the request to each stratum 1, because otherwise every request would be cached already and respond equally fast. 

That might work if there were only one client probing through each squid at a time, but collapsed forwarding would still be taking place in the squid, so if a request to a slow stratum 1 from one client was just finishing up when a second client makes the same request, the second client would get a misleading fast response from the slow stratum 1 and incorrectly choose it as its best stratum 1.

Also, Cloudflare ignores `Cache-Control` headers from the client, so the scheme wouldn't work through that CDN to tell which stratum 1 is closer.

Perhaps a way around both of those problems would be to instead read an API on the stratum 1s that ignores a parameter, and make sure every client inserts a unique string as the parameter, a strong hash of some kind.  For Cloudflare it would also need to insert a different string each time it does a mount or it might get cached responses from a previous mount. Those would then clutter up the caches but maybe they would be small enough and mounts are infrequent enough that they won't push too much else out of the caches.

**Comment by DrDaveD** [comment]
Or maybe if they did an http POST instead of a GET there wouldn't be any issue with cluttering caches.

**Comment by rptaylor** [comment]
> Yes CONNECT is required to access an https page through a forward proxy. However it cannot cache the response. Every https response will be different and only understandable to the client & server that know the encryption key for the connection. CONNECT only passes bytes through and does not understand them. That would be a man in the middle attack if it could.

Right, so to further clarify, I think [this](http://libtorrent.org/reference-Settings.html#enum-proxy-type-t) is saying "The server is assumed to be an HTTP proxy. If the transport used for the connection is HTTPS, the server is assumed to support the [CONNECT](http://tools.ietf.org/html/draft-luotonen-web-proxy-tunneling-01) method, otherwise for HTTP transport the CONNECT method is not required and a plain proxy will suffice."  If this interpretation is correct (to be confirmed), caching forward proxies would work fine as they are.

> I suppose it might potentially work for the clients to do multiple GET requests. It would also be N times the bandwidth, 

Neither of the possibilities (aria2 or libtorrent) discussed here do any bandwidth amplification - that would be inefficient indeed. Downloading a single file is parallelized by fetching individual pieces concurrently from multiple servers, but no individual piece would be fetched more than once unless the downloaded data is corrupt. I think this is implemented via byte range headers in GET requests.

**Comment by DrDaveD** [comment]
> > Yes CONNECT is required to access an https page through a forward proxy. However it cannot cache the response. Every https response will be different and only understandable to the client & server that know the encryption key for the connection. CONNECT only passes bytes through and does not understand them. That would be a man in the middle attack if it could.
> 
> Right, so to further clarify, I think [this](http://libtorrent.org/reference-Settings.html#enum-proxy-type-t) is saying "The server is assumed to be an HTTP proxy. If the transport used for the connection is HTTPS, the server is assumed to support the [CONNECT](http://tools.ietf.org/html/draft-luotonen-web-proxy-tunneling-01) method, otherwise for HTTP transport the CONNECT method is not required and a plain proxy will suffice." If this interpretation is correct (to be confirmed), caching forward proxies would work fine as they are.

OK so you're saying that it would be using http and so it would not use CONNECT.  I was losing the track of the discussion.  You were responding to my initial comment where I objected to the use of CONNECT, but apparently it wouldn't need to use that.

> > I suppose it might potentially work for the clients to do multiple GET requests. It would also be N times the bandwidth,
> 
> Neither of the possibilities (aria2 or libtorrent) discussed here do any bandwidth amplification - that would be inefficient indeed. Downloading a single file is parallelized by fetching individual pieces concurrently from multiple servers, but no individual piece would be fetched more than once unless the downloaded data is corrupt. I think this is implemented via byte range headers in GET requests.

Oh, I thought you were talking about dynamically locating the closest servers.

I don't see the value of reading chunks of larger files from multiple servers when the servers are distributed in small numbers around the world.  We are not limited by bandwidth of the servers, we are limited by the latency of far away servers when we download large numbers of small files.  Torrent is designed for lots of people downloading the same large files, which is not the nature of most of the files in CVMFS.  If we have larger files we'd be better off reading chunks in parallel from the closest stratum 1, I think.

**Comment by rptaylor** [comment]
> OK so you're saying that it would be using http and so it would not use CONNECT.

https://github.com/arvidn/libtorrent/issues/7656  could clarify.

> Oh, I thought you were talking about dynamically locating the closest servers.

The original idea was that it shouldn't matter which server is closest/best if you are using all of them.
But aside from that I agree that dynamically determining the lowest latency one as you suggest could be a good separate improvement that would obviate geoAPI.
 
> I don't see the value of reading chunks of larger files from multiple servers when the servers are distributed in small numbers around the world. We are not limited by bandwidth of the servers, we are limited by the latency of far away servers when we download large numbers of small files. Torrent is designed for lots of people loading the same large files, which is not the nature of most of the files in CVMFS. If we have larger files we'd be better off reading chunks in parallel from the closest stratum 1, I think.

The premise is instead of getting (max bandwidth of best server) you get (max bandwidth of best server + available bandwidth from another server + available bandwidth from another server + .... ). That being said I agree that latency is an equally important and potentially limiting factor (although not all deployments have stratum servers distributed across the world, as opposed to regional or smaller deployments, so how much of a limitation latency is may vary), and I think you are right it would work better for large files than small chunks.  But chunk size is configurable and accelerating large data files on CVMFS would be valuable. But the bigger game changer would be a hypothetical "phase 2" implementation where clients could swarm to torrent data amongst themselves in a local private network, where the low latency should provide performance gains for any size of file.

To try to briefly summarize the initial investigation:
-  aria2 appears to be a perfect solution for parallelized downloads from stratum servers, but is unfortunately not very well documented, and infrequently updated. Although the metalink standard is not that widely implemented, it is [used](https://en.wikipedia.org/wiki/Metalink) by yum/dnf so it should be safe to rely on.
- It seems libtorrent should also work however this approach would require clients to have finer-grained checksum data. I wonder if this could be handled similarly to [grafting](https://cvmfs.readthedocs.io/en/stable/cpt-repo.html#sct-grafting), where admins need to determine and provide chunk checksums with a separate mechanism. Unlike aria2, adoption of this approach could hypothetically lead to a client "swarm" in a local private network.

Maybe further discussion or investigation of feasibility would be suitable as a workshop topic but it would depend on the developers' assessment.

**Comment by DrDaveD** [comment]
Because we do have so many small files, I don't expect much performance gain from a libtorrent-based implementation.  The main value I see is the possibility of not requiring a local squid service.  On the other hand, I think that it will be a lot harder to debug problems, and that it's likely that many networks would be configured to block torrent-type traffic.  So the benefit of such an effort is very questionable to me.

**Comment by rptaylor** [comment]
Yes, while most frequently used software repos often have small files, in general it depends on the nature of the content that is published, and the configured chunk size. Anyway there was a positive response to https://github.com/arvidn/libtorrent/issues/7656
